d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,583 @@
+/*
+   Copyright (c) 2023 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.d2.xds;
+
+import com.google.common.base.Joiner;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.rpc.Code;
+import indis.XdsD2;
+import io.envoyproxy.envoy.service.discovery.v3.AggregatedDiscoveryServiceGrpc;
+import io.envoyproxy.envoy.service.discovery.v3.DeltaDiscoveryRequest;
+import io.envoyproxy.envoy.service.discovery.v3.DeltaDiscoveryResponse;
+import io.envoyproxy.envoy.service.discovery.v3.Resource;
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.internal.BackoffPolicy;
+import io.grpc.internal.ExponentialBackoffPolicy;
+import io.grpc.stub.StreamObserver;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class XdsClientImpl extends XdsClient
+{
+  private static final Logger _log = LoggerFactory.getLogger(XdsClientImpl.class);
+
+  private final Map<String, ResourceSubscriber> _d2NodeSubscribers = new HashMap<>();
+  private final Map<String, ResourceSubscriber> _d2NodeMapSubscribers = new HashMap<>();
+
+  private final Node _node;
+  private final ManagedChannel _managedChannel;
+  private final ScheduledExecutorService _executorService;
+  private final BackoffPolicy.Provider _backoffPolicyProvider = new ExponentialBackoffPolicy.Provider();
```

Comment:
Let's be explicit about the multiplier and the maximum: https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/ExponentialBackoffPolicy.java#L79 . So that we can fully control our client's behavior.

Reply:
Yeah i agree. But it seems all the setters in this class are package-private (a lot of native gRPC code behaves like this, ugh..). What are you thoughts on copying this class over to our own code base?

Reply:
Oh actually, i think these parameters are part of the documented algorithm here: https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md